File: Quicksort.py
from mpi4py import MPI
import numpy as np
import time
import platform
import statistics

def choose_pivot(arr, start, end):
    """Choose median-of-three pivot to reduce variability."""
    mid = (start + end) // 2
    candidates = [(arr[start], start), (arr[mid], mid), (arr[end], end)]
    candidates.sort()
    return candidates[1][1]  # Median index

def sequential_quicksort(arr, start, end):
    """Sequential Quicksort with median-of-three pivot."""
    if start < end:
        pivot_idx = choose_pivot(arr, start, end)
        arr[pivot_idx], arr[end] = arr[end], arr[pivot_idx]
        pivot = arr[end]
        i = start - 1
        for j in range(start, end + 1):
            if arr[j] <= pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        pi = i
        sequential_quicksort(arr, start, pi - 1)
        sequential_quicksort(arr, pi + 1, end)

def merge_sorted_arrays(arr1, arr2):
    """Merge two sorted arrays."""
    result = []
    i, j = 0, 0
    while i < len(arr1) and j < len(arr2):
        if arr1[i] <= arr2[j]:
            result.append(arr1[i])
            i += 1
        else:
            result.append(arr2[j])
            j += 1
    result.extend(arr1[i:])
    result.extend(arr2[j:])
    return result

def parallel_quicksort(comm, rank, size, data, n):
    """Parallel Quicksort with accurate timing."""
    if size == 1:
        start_comp = time.perf_counter()
        sequential_quicksort(data, 0, n - 1)  # Sort in-place, no copy
        comp_time = time.perf_counter() - start_comp
        return data.tolist(), comp_time, 0.0

    chunk_size = n // size
    if rank == size - 1:
        chunk_size = n - (size - 1) * chunk_size

    local_data = np.zeros(chunk_size, dtype=np.int32)
    
    comm_time = 0
    start_comm = time.perf_counter()
    comm.Scatter(data, local_data, root=0)
    comm_time += time.perf_counter() - start_comm

    start_comp = time.perf_counter()
    sequential_quicksort(local_data, 0, len(local_data) - 1)
    comp_time = time.perf_counter() - start_comp

    start_comm = time.perf_counter()
    if rank == 0:
        sorted_chunks = [local_data]
        for i in range(1, size):
            recv_chunk = np.zeros(n // size if i < size - 1 else n - (size - 1) * (n // size), dtype=np.int32)
            comm.Recv(recv_chunk, source=i)
            sorted_chunks.append(recv_chunk)
        result = sorted_chunks[0].tolist()
        for i in range(1, len(sorted_chunks)):
            result = merge_sorted_arrays(result, sorted_chunks[i].tolist())
    else:
        comm.Send(local_data, dest=0)
        result = None
    comm_time += time.perf_counter() - start_comm

    return result, comp_time, comm_time

def main():
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    if rank == 0:
        print("=" * 50)
        print("MPI Quicksort Performance Test".center(50))
        print("=" * 50)
        print(f"System : {platform.processor()}")
        print(f"OS     : {platform.system()} {platform.release()}")
        print(f"MPI    : {MPI.Get_library_version().splitlines()[0]}")
        print(f"Processes: {size}")
        print("-" * 50)

    num_iterations = 20  # Increased for reliability

    if rank == 0:
        try:
            num_sizes = int(input("Enter the number of different array sizes to test: "))
            if num_sizes <= 0:
                print("Error: Please enter a positive number.")
                comm.Abort()
            n_values = []
            for _ in range(num_sizes):
                n = int(input("Enter an array size: "))
                if n <= 0:
                    print("Error: Please enter a positive number.")
                    comm.Abort()
                n_values.append(n)
        except ValueError:
            print("Error: Array size must be an integer.")
            comm.Abort()
    else:
        num_sizes = None
        n_values = None

    num_sizes = comm.bcast(num_sizes, root=0)
    if rank == 0:
        n_values = n_values
    else:
        n_values = [None] * num_sizes
    n_values = comm.bcast(n_values, root=0)

    for n in n_values:
        if rank == 0:
            try:
                print(f"\nTest Case: Array Size = {n}")
                print(f"Enter {n} space-separated integers:")
                array_input = input().strip().split()
                if len(array_input) != n:
                    print(f"Error: Expected {n} elements, got {len(array_input)}.")
                    comm.Abort()
                data = [int(x) for x in array_input]
                data = np.array(data, dtype=np.int32)
                if n <= 11:
                    print(f"Input Array: {data.tolist()}")
                else:
                    print(f"Input Array Size: {n}")
            except ValueError:
                print("Error: All array elements must be integers.")
                comm.Abort()
        else:
            data = None

        if rank == 0:
            data = np.array(data, dtype=np.int32)
        else:
            data = np.zeros(n, dtype=np.int32)
        comm.Bcast(data, root=0)

        seq_times = []
        seq_sorted_data = None
        if rank == 0:
            for _ in range(num_iterations):
                seq_data = data.copy()
                start_time = time.perf_counter()
                sequential_quicksort(seq_data, 0, n - 1)
                seq_times.append(time.perf_counter() - start_time)
                if not np.all(np.diff(seq_data) >= 0):
                    print("Error: Sequential sort failed!")
                seq_sorted_data = seq_data
            avg_seq_time = statistics.mean(seq_times)
            std_seq_time = statistics.stdev(seq_times) if num_iterations > 1 else 0
            print("\nSequential Quicksort Results:")
            print(f"  Time (avg ± std): {avg_seq_time*1e6:>8.2f} ± {std_seq_time*1e6:>6.2f} μs")
            if n <= 11:
                print(f"  Sorted Array    : {seq_sorted_data.tolist()}")

        par_times = []
        comp_times = []
        comm_times = []
        for _ in range(num_iterations):
            comm.Barrier()
            start_time = time.perf_counter()
            sorted_data, comp_time, comm_time = parallel_quicksort(comm, rank, size, data, n)
            comm.Barrier()
            par_times.append(time.perf_counter() - start_time)
            comp_times.append(comp_time)
            comm_times.append(comm_time)

        all_comp_times = comm.gather(comp_times, root=0)

        if rank == 0:
            avg_par_time = statistics.mean(par_times)
            std_par_time = statistics.stdev(par_times) if num_iterations > 1 else 0
            avg_comm_time = statistics.mean(comm_times)
            avg_comp_time = statistics.mean(comp_times)
            overhead_time = avg_par_time - (avg_comm_time + avg_comp_time)
            
            print("\nParallel Quicksort Results:")
            print(f"  Time (avg ± std): {avg_par_time*1e6:>8.2f} ± {std_par_time*1e6:>6.2f} μs ({size} processes)")
            print(f"  Communication   : {avg_comm_time*1e6:>8.2f} μs ({avg_comm_time/avg_par_time*100:>5.1f}%)")
            print(f"  Computation     : {avg_comp_time*1e6:>8.2f} μs ({avg_comp_time/avg_par_time*100:>5.1f}%)")
            print(f"  Overhead        : {overhead_time*1e6:>8.2f} μs ({overhead_time/avg_par_time*100:>5.1f}%)")
            
            flat_comp_times = [t for iter_times in all_comp_times for t in iter_times]
            max_comp = max(flat_comp_times)
            min_comp = min(flat_comp_times)
            avg_comp = statistics.mean(flat_comp_times)
            print(f"  Load Balance    : Max {max_comp*1e6:>6.2f} μs, Min {min_comp*1e6:>6.2f} μs, Avg {avg_comp*1e6:>6.2f} μs")
            
            if sorted_data is not None and not np.all(np.diff(np.array(sorted_data)) >= 0):
                print("Error: Parallel sort failed!")
            if n <= 11:
                print(f"  Sorted Array    : {sorted_data}")
            
            if avg_par_time > 0:
                speedup = avg_seq_time / avg_par_time
                efficiency = speedup / size
                print(f"  Speedup         : {speedup:>8.2f}x")
                print(f"  Efficiency      : {efficiency:>8.2f}")
            else:
                print("  Speedup         : Not calculated (zero parallel time)")
            
            print("=" * 50)

if __name__ == "__main__":
    main()

File: Quicksort.cpp
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <time.h>

#define ARRAY_SIZE 10000 // Size of the array to sort

// Function to swap two elements
void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

// Function to select a better pivot (median-of-three)
int choose_pivot(int arr[], int low, int high) {
    int mid = low + (high - low) / 2;
    // Sort low, mid, high values to get median
    if (arr[low] > arr[mid]) swap(&arr[low], &arr[mid]);
    if (arr[low] > arr[high]) swap(&arr[low], &arr[high]);
    if (arr[mid] > arr[high]) swap(&arr[mid], &arr[high]);
    // Place pivot at high
    swap(&arr[mid], &arr[high]);
    return arr[high];
}

// Function to partition the array
int partition(int arr[], int low, int high) {
    int pivot = choose_pivot(arr, low, high);
    int i = low - 1;

    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return i + 1;
}

// Sequential Quicksort with loop prevention
void quicksort(int arr[], int low, int high) {
    if (low < high && low >= 0 && high < ARRAY_SIZE) { // Prevent invalid indices
        int pi = partition(arr, low, high);
        quicksort(arr, low, pi - 1);
        quicksort(arr, pi + 1, high);
    }
}

// Function to merge sorted subarrays
void merge(int arr[], int temp[], int low, int mid, int high) {
    int i = low, j = mid + 1, k = low;

    while (i <= mid && j <= high) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }

    while (i <= mid) {
        temp[k++] = arr[i++];
    }

    while (j <= high) {
        temp[k++] = arr[j++];
    }

    for (i = low; i <= high; i++) {
        arr[i] = temp[i];
    }
}

int main(int argc, char* argv[]) {
    int rank, size;
    int *arr = NULL, *local_arr = NULL, *temp = NULL;
    int local_size, *sendcounts = NULL, *displs = NULL;
    double start_time, end_time;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    // Check if ARRAY_SIZE is valid
    if (ARRAY_SIZE <= 0) {
        if (rank == 0) {
            fprintf(stderr, "Error: ARRAY_SIZE must be positive\n");
        }
        MPI_Finalize();
        return 1;
    }

    // Allocate sendcounts and displacements for uneven distribution
    sendcounts = (int*)malloc(size * sizeof(int));
    displs = (int*)malloc(size * sizeof(int));
    local_size = ARRAY_SIZE / size;
    int remainder = ARRAY_SIZE % size;

    for (int i = 0; i < size; i++) {
        sendcounts[i] = (i < remainder) ? local_size + 1 : local_size;
        displs[i] = (i == 0) ? 0 : displs[i - 1] + sendcounts[i - 1];
    }

    // Root process initializes the array
    if (rank == 0) {
        arr = (int*)malloc(ARRAY_SIZE * sizeof(int));
        srand(time(NULL));
        for (int i = 0; i < ARRAY_SIZE; i++) {
            arr[i] = rand() % 10000;
        }
        printf("Initial array (first 10 elements): ");
        for (int i = 0; i < 10 && i < ARRAY_SIZE; i++) {
            printf("%d ", arr[i]);
        }
        printf("\n");
    }

    // Allocate local array for each process
    local_arr = (int*)malloc(sendcounts[rank] * sizeof(int));
    temp = (int*)malloc(ARRAY_SIZE * sizeof(int));

    // Start timing
    start_time = MPI_Wtime();

    // Scatter the array to all processes
    MPI_Scatterv(arr, sendcounts, displs, MPI_INT, local_arr, sendcounts[rank], MPI_INT, 0, MPI_COMM_WORLD);

    // Each process sorts its local array
    quicksort(local_arr, 0, sendcounts[rank] - 1);

    // Gather sorted subarrays back to root
    MPI_Gatherv(local_arr, sendcounts[rank], MPI_INT, arr, sendcounts, displs, MPI_INT, 0, MPI_COMM_WORLD);

    // Root process merges the sorted subarrays
    if (rank == 0) {
        // Copy gathered array to temp for merging
        for (int i = 0; i < ARRAY_SIZE; i++) {
            temp[i] = arr[i];
        }

        // Perform merge for all subarrays
        for (int step = 1; step < size; step *= 2) {
            for (int i = 0; i < size; i += 2 * step) {
                int low = displs[i];
                int mid = (i + step < size) ? displs[i + step] - 1 : ARRAY_SIZE - 1;
                int high = (i + 2 * step < size) ? displs[i + 2 * step] - 1 : ARRAY_SIZE - 1;

                if (mid >= ARRAY_SIZE || low >= high) continue;
                if (high >= ARRAY_SIZE) high = ARRAY_SIZE - 1;

                merge(arr, temp, low, mid, high);
            }
        }

        // End timing
        end_time = MPI_Wtime();

        // Print first 10 elements of sorted array
        printf("Sorted array (first 10 elements): ");
        for (int i = 0; i < 10 && i < ARRAY_SIZE; i++) {
            printf("%d ", arr[i]);
        }
        printf("\n");

        // Verify if array is sorted
        int is_sorted = 1;
        for (int i = 1; i < ARRAY_SIZE; i++) {
            if (arr[i - 1] > arr[i]) {
                is_sorted = 0;
                break;
            }
        }
        printf("Array is %s\n", is_sorted ? "sorted" : "not sorted");
        printf("Time taken: %f seconds\n", end_time - start_time);
    }

    // Clean up
    if (rank == 0) {
        free(arr);
    }
    free(local_arr);
    free(temp);
    free(sendcounts);
    free(displs);

    MPI_Finalize();
    return 0;
}
